pkg/deploy/controller/deploymentconfig/controller.go

```diff
@@ -54,230 +58,191 @@ func (e transientError) Error() string {
-	}
-
-	// check to see if there are inflight deployments
-	if inflightDeployment != nil {
+	// Wait for deployment cancellations before reconciling or creating a new
+	// deployment to avoid competing with existing deployment processes.
+	if awaitingCancellations {
+		c.recorder.Eventf(config, "DeploymentAwaitingCancellation", "Deployment of version %d awaiting cancellation of older running deployments", config.LatestVersion)
 		// raise a transientError so that the deployment config can be re-queued
-		glog.V(4).Infof("Found previous inflight deployment for %s - will requeue", deployutil.LabelForDeploymentConfig(config))
 		return transientError(fmt.Sprintf("found previous inflight deployment for %s - requeuing", deployutil.LabelForDeploymentConfig(config)))
 	}
-
-	// Try and build a deployment for the config.
-	deployment, err := c.makeDeployment(config)
+	// If the latest deployment already exists, reconcile existing deployments
+	// and return early.
+	if latestDeploymentExists {
+		// If the latest deployment is still running, try again later. We don't
+		// want to compete with the deployer.
+		if !deployutil.IsTerminatedDeployment(latestDeployment) {
+			return nil
+		}
+		return c.reconcileExistingDeployments(existingDeployments, config)
+	}
+	// No deployments are running and the latest deployment doesn't exist, so
+	// create the new deployment.
+	deployment, err := deployutil.MakeDeployment(config, c.codec)
 	if err != nil {
 		return fatalError(fmt.Sprintf("couldn't make deployment from (potentially invalid) deployment config %s: %v", deployutil.LabelForDeploymentConfig(config), err))
 	}
-
-	// Compute the desired replicas for the deployment. Use the last completed
-	// deployment's current replica count, or the config template if there is no
-	// prior completed deployment available.
-	desiredReplicas := config.Template.ControllerTemplate.Replicas
-	if len(existingDeployments.Items) > 0 {
-		sort.Sort(deployutil.ByLatestVersionDesc(existingDeployments.Items))
-		for _, existing := range existingDeployments.Items {
-			if deployutil.DeploymentStatusFor(&existing) == deployapi.DeploymentStatusComplete {
-				desiredReplicas = existing.Spec.Replicas
-				glog.V(4).Infof("Desired replicas for %s set to %d based on prior completed deployment %s", deployutil.LabelForDeploymentConfig(config), desiredReplicas, existing.Name)
-				break
-			}
-		}
-	}
-	deployment.Annotations[deployapi.DesiredReplicasAnnotation] = strconv.Itoa(desiredReplicas)
-
-	// Create the deployment.
-	if _, err := c.deploymentClient.createDeployment(config.Namespace, deployment); err == nil {
-		glog.V(4).Infof("Created deployment for deployment config %s", deployutil.LabelForDeploymentConfig(config))
-		return nil
-	} else {
-		// If the deployment was already created, just move on. The cache could be stale, or another
-		// process could have already handled this update.
+	created, err := c.kubeClient.ReplicationControllers(config.Namespace).Create(deployment)
```

Comment:
> @liggitt tiebreaker? Should MakeDeployment(config) return a replicationController with namespace=config.Namespace, given the resource is fed to a namespacer?
> 
> I don't feel strongly, just want to keep style consistent.

Yes.  Eliminate all doubt if the object gets disconnect and get smacked by the API server if something insane happens and it breaks.  It also makes any object output in a log or trace unambiguous.